src/java.base/share/classes/java/io/BufferedReader.java

```diff
@@ -99,7 +98,7 @@ public class BufferedReader extends Reader {
      * @throws IllegalArgumentException  If {@code sz <= 0}
      */
     public BufferedReader(Reader in, int sz) {
-        super(in);
+        Objects.requireNonNull(in);
```

Comment:
Not sure if that even matters - but there will be a slight change of behaviour here if `InternalLock.CAN_USE_INTERNAL_LOCK` is ever `false`. Instead of synchronizing on `in`, the `BufferedReader` will synchronize on `this`.
Now that I think of it - it probably does matter since even if CAN_USE_INTERNAL_LOCK is true,  untrusted subclasses of BufferedReader calling this constructor might expect the locking to be performed on `in`?

Reply:
> Not sure if that even matters - but there will be a slight change of behaviour here if `InternalLock.CAN_USE_INTERNAL_LOCK` is ever `false`. Instead of synchronizing on `in`, the `BufferedReader` will synchronize on `this`.

Good!  We can change this so that depends on whether BufferedReader is extended and whether the given Reader is trusted. It's not clear if anyone could reliably depend on undocumented behavior like this but we have to be cautious at the same time.

Reply:
Thanks - the same issue appears with `BufferedWriter`/`Writer`.

Reply:
The solution is to add the `private` constructor:
```java
	private Reader(Object fallbackLock, Void internal) {
		// assert fallbackLock != null;
		// use InternalLock for trusted classes
		Class<?> clazz = getClass();
		if (clazz == InputStreamReader.class
			|| clazz == BufferedReader.class
			|| clazz == FileReader.class
			|| clazz == sun.nio.cs.StreamDecoder.class) {
			this.lock = InternalLock.newLockOr(fallbackLock);
		} else {
			this.lock = fallbackLock;
		}
	}
```
to `Reader` (and an equivalent `private` constructor to `Writer`).

---

The `protected` `Reader` constructors can then be changed to:
```java
	protected Reader() {
		this(this, null);
	}
```
and
```java
	protected Reader(Object lock) {
		this(Objects.requireNonNull(lock), null);
	}
```

---

Doing so will allow this change to be reverted:
```suggestion
        super(in);
```

Reply:
> Thanks - the same issue appears with `BufferedWriter`/`Writer`.

Right. I think we can reduce this to the case where a BufferedReader is not subclassed, and it wraps an InputStreamReader or FileReader that have not been subclassed. Similarly the case where a BufferedWriter is not subclassed and it wraps an OutputStreamWriter or FileWriter that have not been subclassed.